LET (from `FROM` clauses) implementation - Parsing and ExprNode conversions

alancai98 opened the pull request:
Part 1 of implementation for `LET` clause (#179) as defined in the high-level [spec](https://github.com/partiql/partiql-spec/issues/15).

Currently just implementing `LET` following `FROM` clauses as they are what's most clearly defined in the high-level spec and use cases.

Future supported clauses will be added later (especially because of #281, _Incorrect execution order for_ `LIMIT` _clause during query execution_)


Current status relating to `LET` following `FROM` clauses (from_let):
- [x] PIG domain model of `LET` clauses (separate [PR](https://github.com/partiql/partiql-lang-kotlin/pull/278))
- [x] `ExprNode` representation of `LET` clauses
- [x] Conversion between `PartiqlAst` and `ExprNode` for `LET` clauses
- [x] V0 Ast Serliazation for `LET` clauses
- [x] Parsing of `LET` clauses
- [x] `LET` parsing tests
(Evaluation to be done in a separate PR before merging into `from-let-final` feature branch)

Possible additions:
- test conversion between `PartiqlAst` and `ExprNode`
- test V0 ast (de)serializer outputs intended error when parsing `LET` clauses

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dlurton commented on lang/src/org/partiql/lang/ast/StatementToExprNode.kt:
```diff
@@ -232,6 +234,18 @@ private class StatementTransformer(val ion: IonSystem) {
+            this.letBindings.map {
+                LetBinding(
+                    it.expr.toExprNode(),
+                    it.name.toSymbolicName()!!
```
Pretty sure that `!!` is not needed here.

alancai98 replied:
Removed the not-null assertion op. Revised `toSymbolicName` to not use nullable guards (`?`)

dlurton commented on lang/test/org/partiql/lang/syntax/SqlParserTest.kt:
```diff
@@ -3092,4 +3096,96 @@ class SqlParserTest : SqlParserTestBase() {
+        select(
+            project = projectX,
+            from = scan(id("table1")),
+            fromLet = let(
+                letBinding(
+                    lit(ionInt(1)),
+                    "A")
+            )
```
Nit:  put all on one line.  same for others below.

alancai98 replied:
Revised formatting

dlurton commented:
Off to a good start--evaluation might be the interesting part.

alancai98 commented:
Updated to resolve David's comments. 

Changed this PR to focus on just parsing and `ExprNode` conversion for `LET`. The evaluation-side implementation of `LET` for `FROM` clauses will have a separate PR. Both of these branches will be merged into the `from-let-clause` base branch before merging into master.

therapon commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -1815,6 +1845,48 @@ class SqlParser(private val ion: IonSystem) : Parser {
+        rem = rem.tail
+
+        if (rem.head?.type?.isIdentifier() != true) {
+            rem.head.err("Expected identifier for $AS-alias", PARSE_EXPECTED_IDENT_FOR_ALIAS)
```
nit: This error message uses `$AS` to interpolate, can we do the same for `LET` on line 1855? 
I do not mind either way as long as we are consistent

alancai98 replied:
Sure, also added to the other error message in the while loop.

therapon reviewed (approved):
minor nit, otherwise looks good

dlurton commented on lang/src/org/partiql/lang/syntax/SqlParser.kt:
```diff
@@ -987,6 +1010,7 @@ class SqlParser(private val ion: IonSystem) : Parser {
             "substring" -> tail.parseSubstring(head!!)
             "trim" -> tail.parseTrim(head!!)
             "extract" -> tail.parseExtract(head!!)
+            "let" -> tail.parseLet()
```
I just bumped into this by accident--this needs to be removed.  It's not helping anything since there are not test cases that cover this.  This might be left over from when we considering allowing `LET` to be a first-class expression.  In any case, this should be removed here...

alancai98 replied:
Will remove when merging `from-let-final` into `master`.

alancai98 replied:
Removed in this [commit](https://github.com/partiql/partiql-lang-kotlin/pull/303/commits/07b654aca2c75e4fb57dce77712b93a9509b7d6e).

dlurton reviewed (commented):
LGTM. Don't forget to squash when merging.  Thanks.
